In [39]:
import pandas as pd

In [69]:
df = pd.read_csv('spam.csv', encoding='latin1')
df = df.iloc[:, :2]
df = df.rename(columns={"v1": "target", "v2": "message"})

In [71]:
df.target.value_counts()

target
ham     4825
spam     747
Name: count, dtype: int64

In [72]:
df['target'] = df.target.map({"ham": 0, "spam": 1}) # label encoding

In [74]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5572 entries, 0 to 5571
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   target   5572 non-null   int64 
 1   message  5572 non-null   object
dtypes: int64(1), object(1)
memory usage: 87.2+ KB


In [96]:
from sklearn.feature_extraction.text import TfidfVectorizer

vectorizer = TfidfVectorizer(max_features=5000)  # Use top 5000 words
X = vectorizer.fit_transform(df['message'])
y = df.target

In [97]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y) # split training vs test sets

In [98]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [101]:
model = LogisticRegression(random_state=42) # initialize model
model.fit(X_train, y_train) # train model

LogisticRegression(random_state=42)

In [102]:
pred = model.predict(X_test)
accuracy_score(y_true=y_test, y_pred=pred)

0.968609865470852

#### Our logistic regression model did well—an accuracy of `97%` is great!